File: src/main/java/io/github/shirohoo/library/controller/user/UserController.java
package io.github.shirohoo.library.controller.user;

import io.github.shirohoo.library.domain.user.User;
import io.github.shirohoo.library.dto.user.UserCreateRequest;
import io.github.shirohoo.library.dto.user.UserResponse;
import io.github.shirohoo.library.dto.user.UserUpdateRequest;
import io.github.shirohoo.library.service.user.UserService;
import java.util.List;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/users")
public class UserController {
    private final UserService userService;

    public UserController(UserService userService) {
        this.userService = userService;
    }

    @GetMapping
    public List<UserResponse> getUsers() {
        return userService.getUsers()
                .stream()
                .map(UserResponse::new)
                .toList();
    }

    @PostMapping
    public void saveUser(@RequestBody UserCreateRequest request) {
        String username = request.getName();
        Integer age = request.getAge();
        userService.saveUser(new User(username, age));
    }

    @PutMapping
    public void updateUserName(@RequestBody UserUpdateRequest request) {
        long userId = request.getId();
        String username = request.getName();
        userService.updateUserName(userId, username);
    }

    @DeleteMapping
    public void deleteUser(@RequestParam String name) {
        userService.deleteUser(name);
    }
}

File: src/main/kotlin/io/github/shirohoo/library/application/controller/user/UserController.kt
package io.github.shirohoo.library.application.controller.user

import io.github.shirohoo.library.application.data.user.*
import io.github.shirohoo.library.application.service.user.UserService
import io.github.shirohoo.library.domain.user.User
import io.github.shirohoo.library.domain.user.UserLoanHistory
import io.github.shirohoo.library.domain.user.UserLoanStatus
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/user")
class UserController(
    private val userService: UserService
) {
    @GetMapping
    fun getUsers(): List<UserResponse> {
        return userService.getUsers()
            .map(::UserResponse)
    }

    @PostMapping
    fun saveUser(@RequestBody request: UserCreateRequest) {
        val user = request.let { User(it.name, it.age) }
        userService.saveUser(user)
    }

    @PutMapping
    fun updateUsername(@RequestBody request: UserUpdateRequest) {
        val (userId, username) = request
        userService.updateUsername(userId, username)
    }

    @DeleteMapping
    fun deleteUser(@RequestBody username: String) = userService.deleteUser(username)

    @GetMapping("/loan")
    fun getUserLoanHistories(): List<UserLoanHistoryResponse> {
        return userService.getUsers()
            .map(::userLoanHistoryResponse)
    }

    private fun userLoanHistoryResponse(user: User) = UserLoanHistoryResponse(
        name = user.name,
        books = user.userLoanHistories.map(::bookHistoryResponse)
    )

    private fun bookHistoryResponse(history: UserLoanHistory) = BookHistoryResponse(
        name = history.bookTitle,
        isReturn = history.status == UserLoanStatus.RETURNED
    )
}